### ISSUE TITLE
Issue working with Teal source (`tl.tl`)

### ISSUE THREAD

**Issue by FourierTransformer** [teal-language/teal-language-server#23]
It currently fails with a syntax error with the following.

```teal
   local function patch_visitors(my_visit_node: Visitor<TypeChecker, NodeKind, Node, Type>,
                                 after_node: VisitorAfterPatcher<TypeChecker, Node, Type>,
                                 my_visit_type?: Visitor<TypeChecker, TypeName, Type, Type>,
                                 after_type?: VisitorAfterPatcher<TypeChecker, Type, Type>):
                                 Visitor<TypeChecker, NodeKind, Node, Type>,
                                 Visitor<TypeChecker, TypeName, Type, Type>
```

It can be fixed by modifying it to:
```teal
   local function patch_visitors(my_visit_node: Visitor<TypeChecker, NodeKind, Node, Type>,
                                 after_node: VisitorAfterPatcher<TypeChecker, Node, Type>,
                                 my_visit_type?: Visitor<TypeChecker, TypeName, Type, Type>,
                                 after_type?: VisitorAfterPatcher<TypeChecker, Type, Type>): Visitor<TypeChecker, NodeKind, Node, Type>, Visitor<TypeChecker, TypeName, Type, Type>
```

_Originally posted by @FourierTransformer in https://github.com/teal-language/teal-language-server/issues/19#issuecomment-2456119992_

I looked into it a bit and it's the `_strip_trailing_colons` in `document.tl` that causes this. Removing the call fixes the issue. 

The comment indicates that it's to help with completion, but I don't think I have my text editor config setup properly, since I can't quite get completion to work. I would be in favor of removing the call to `_strip_trailing_colons` since it cab  generate false positives for users, and then reworking completion to work without it.

**Comment by FourierTransformer** [comment]
I ended up getting code completion working. It turned out my editor didn't send a `triggerKind` variable. Once I figured that out (and did a nil check) it was smooth sailing, and all the completion stuff just started working! I even updated it with `kind` so there are some icons and got function parameter completion going.

So yeah, I don't think the `_strip_trailing_colons` was still needed. I need to clean up my branch, but I'll update my PR (#22) with the changes I've been playing with.

**Comment by svermeulen** [comment]
Thanks for the help.  Yes, agreed, would be great to get rid of that horrible hack

**Comment by FourierTransformer** [comment]
Alrighty, before that PR gets too big, I figured I should push it up. It's decently stable and I've made note on things that are sorta broken. I took a look at some other LSP implementations, and saw how they handled completion of function params, and may try to emulate those (I specifically liked how [basedpyright](https://github.com/DetachHead/basedpyright) did it). I initially had them just being injected, but that was rather annoying due to optionals and polymorphism .
